### DISCUSSION TITLE
how can i play sound for each detection

### DISCUSSION THREAD

**Discussion by 7assanx7** [ultralytics/ultralytics#1814] [Q&A]
`# Ultralytics YOLO 🚀, GPL-3.0 license

import torch

from ultralytics.yolo.engine.predictor import BasePredictor
from ultralytics.yolo.engine.results import Results
from ultralytics.yolo.utils import DEFAULT_CFG, ROOT, ops
from ultralytics.yolo.utils.plotting import Annotator, colors, save_one_box

class DetectionPredictor(BasePredictor):

    def get_annotator(self, img):
        return Annotator(img, line_width=self.args.line_thickness, example=str(self.model.names))

    def preprocess(self, img):
        img = (img if isinstance(img, torch.Tensor) else torch.from_numpy(img)).to(self.model.device)
        img = img.half() if self.model.fp16 else img.float()  # uint8 to fp16/32
        img /= 255  # 0 - 255 to 0.0 - 1.0
        return img

    def postprocess(self, preds, img, orig_imgs):
        preds = ops.non_max_suppression(preds,
                                        self.args.conf,
                                        self.args.iou,
                                        agnostic=self.args.agnostic_nms,
                                        max_det=self.args.max_det,
                                        classes=self.args.classes)

        results = []
        for i, pred in enumerate(preds):
            orig_img = orig_imgs[i] if isinstance(orig_imgs, list) else orig_imgs
            if not isinstance(orig_imgs, torch.Tensor):
                pred[:, :4] = ops.scale_boxes(img.shape[2:], pred[:, :4], orig_img.shape)
            path, _, _, _, _ = self.batch
            img_path = path[i] if isinstance(path, list) else path
            results.append(Results(orig_img=orig_img, path=img_path, names=self.model.names, boxes=pred))
        return results

    def write_results(self, idx, results, batch):
        p, im, im0 = batch
        log_string = ''
        if len(im.shape) == 3:
            im = im[None]  # expand for batch dim
        self.seen += 1
        imc = im0.copy() if self.args.save_crop else im0
        if self.source_type.webcam or self.source_type.from_img:  # batch_size >= 1
            log_string += f'{idx}: '
            frame = self.dataset.count
        else:
            frame = getattr(self.dataset, 'frame', 0)
        self.data_path = p
        self.txt_path = str(self.save_dir / 'labels' / p.stem) + ('' if self.dataset.mode == 'image' else f'_{frame}')
        log_string += '%gx%g ' % im.shape[2:]  # print string
        self.annotator = self.get_annotator(im0)

        det = results[idx].boxes  # TODO: make boxes inherit from tensors
        if len(det) == 0:
            return f'{log_string}(no detections), '
        for c in det.cls.unique():
            n = (det.cls == c).sum()  # detections per class
            log_string += f"{n} {self.model.names[int(c)]}{'s' * (n > 1)}, "

        # write
        for d in reversed(det):
            c, conf, id = int(d.cls), float(d.conf), None if d.id is None else int(d.id.item())
            if self.args.save_txt:  # Write to file
                line = (c, *d.xywhn.view(-1)) + (conf, ) * self.args.save_conf + (() if id is None else (id, ))
                with open(f'{self.txt_path}.txt', 'a') as f:
                    f.write(('%g ' * len(line)).rstrip() % line + '\n')
            if self.args.save or self.args.show:  # Add bbox to image
                name = ('' if id is None else f'id:{id} ') + self.model.names[c]
                label = None if self.args.hide_labels else (name if self.args.hide_conf else f'{name} {conf:.2f}')
                self.annotator.box_label(d.xyxy.squeeze(), label, color=colors(c, True))
            if self.args.save_crop:
                save_one_box(d.xyxy,
                             imc,
                             file=self.save_dir / 'crops' / self.model.names[c] / f'{self.data_path.stem}.jpg',
                             BGR=True)

        return log_string

def predict(cfg=DEFAULT_CFG, use_python=False):
    model = cfg.model or 'yolov8n.pt'
    source = cfg.source if cfg.source is not None else ROOT / 'assets' if (ROOT / 'assets').exists() \
        else 'https://ultralytics.com/images/bus.jpg'

    args = dict(model=model, source=source)
    if use_python:
        from ultralytics import YOLO
        YOLO(model)(**args)
    else:
        predictor = DetectionPredictor(overrides=args)
        predictor.predict_cli()

if __name__ == '__main__':
    predict()`

how can i play sound for each detection or class in real time detection
i want to if class 0 detected play sound by mp3 file

**Comment by dh031200** [comment]
@7assanx7  Using the generator as shown below, you can use yolo frame by frame.
Set `stram=True` for generator.
Use r.boxes to get the data for each frame.


```python
from ultralytics import YOLO

model = YOLO('yolov8n.pt')
results = model.predict(source=src, verbose=False, stream=True)
for r in results:
    if 0 in r.boxes.cls:    # boxes object contains bbox outputs.
        # YOUR CODE
        # play sound function
```
-------
This is sample  of r.boxes:
```
boxes: tensor([[ 1.2615e+03,  5.8268e+02,  1.3865e+03,  8.5076e+02,  1.0000e+00,  9.1142e-01,  1.5000e+01],
        [ 1.8397e+02,  7.3267e+02,  4.2540e+02,  8.5830e+02,  2.0000e+00,  8.9236e-01,  5.0000e+00],
        [-2.7735e-01,  7.4888e+02,  8.5042e+01,  9.0250e+02,  3.0000e+00,  7.8997e-01,  5.0000e+00],
        [ 6.6940e+02,  1.9053e+00,  7.5318e+02,  7.5838e+02,  4.0000e+00,  7.3303e-01,  1.7000e+01],
        [ 6.0532e+02,  5.6310e+02,  7.0006e+02,  8.0803e+02,  5.0000e+00,  7.5743e-01,  2.9000e+01]])
cls: tensor([15.,  5.,  5., 17., 29.])
conf: tensor([0.9114, 0.8924, 0.7900, 0.7330, 0.7574])
data: tensor([[ 1.2615e+03,  5.8268e+02,  1.3865e+03,  8.5076e+02,  1.0000e+00,  9.1142e-01,  1.5000e+01],
        [ 1.8397e+02,  7.3267e+02,  4.2540e+02,  8.5830e+02,  2.0000e+00,  8.9236e-01,  5.0000e+00],
        [-2.7735e-01,  7.4888e+02,  8.5042e+01,  9.0250e+02,  3.0000e+00,  7.8997e-01,  5.0000e+00],
        [ 6.6940e+02,  1.9053e+00,  7.5318e+02,  7.5838e+02,  4.0000e+00,  7.3303e-01,  1.7000e+01],
        [ 6.0532e+02,  5.6310e+02,  7.0006e+02,  8.0803e+02,  5.0000e+00,  7.5743e-01,  2.9000e+01]])
id: tensor([1., 2., 3., 4., 5.])
is_track: True
orig_shape: tensor([1080, 1920])
shape: torch.Size([5, 7])
xywh: tensor([[1324.0040,  716.7162,  125.0505,  268.0803],
        [ 304.6838,  795.4862,  241.4294,  125.6306],
        [  42.3822,  825.6887,   85.3192,  153.6231],
        [ 711.2910,  380.1426,   83.7773,  756.4745],
        [ 652.6879,  685.5687,   94.7351,  244.9300]])
xywhn: tensor([[0.6896, 0.6636, 0.0651, 0.2482],
        [0.1587, 0.7366, 0.1257, 0.1163],
        [0.0221, 0.7645, 0.0444, 0.1422],
        [0.3705, 0.3520, 0.0436, 0.7004],
        [0.3399, 0.6348, 0.0493, 0.2268]])
xyxy: tensor([[ 1.2615e+03,  5.8268e+02,  1.3865e+03,  8.5076e+02],
        [ 1.8397e+02,  7.3267e+02,  4.2540e+02,  8.5830e+02],
        [-2.7735e-01,  7.4888e+02,  8.5042e+01,  9.0250e+02],
        [ 6.6940e+02,  1.9053e+00,  7.5318e+02,  7.5838e+02],
        [ 6.0532e+02,  5.6310e+02,  7.0006e+02,  8.0803e+02]])
xyxyn: tensor([[ 6.5702e-01,  5.3951e-01,  7.2215e-01,  7.8774e-01],
        [ 9.5817e-02,  6.7840e-01,  2.2156e-01,  7.9472e-01],
        [-1.4445e-04,  6.9340e-01,  4.4293e-02,  8.3565e-01],
        [ 3.4865e-01,  1.7642e-03,  3.9228e-01,  7.0220e-01],
        [ 3.1527e-01,  5.2139e-01,  3.6461e-01,  7.4818e-01]])
```

**Reply by 7assanx7** [reply]
thank you it's working but when i'm use gTTS library to speech in real time detection but detection very slow when detect the class that do this function ,do you have idea what is problem ?

**Reply by dh031200** [reply]
How about creating an audio file in advance and then playing that audio in a thread when an object is detected?
As far as I know gTTS takes quite a while to process. This way you don't need to use gTTS.

**Reply by 7assanx7** [reply]
I used this method, but it plays the sound continuously without ending the content of the audio file like this :

from playsound import playsound
import threading

def play_audio():
    playsound("welcome.wav")

for r in results:
    if 0 in r.boxes.cls:
        audio_thread = threading.Thread(target=play_audio)
        audio_thread.start()
        
it work fine without lag in real time detection but as I said it play audio files many times without finishing the sound that make it like echo

**Reply by dh031200** [reply]
As far as I know `pygame` library can solve that problem. `playsound` cannot stop sound, but `pygame` can.
Unfortunately, I don't know the details, I've only used it a few times.

Install pygame:
`pip install pygame`

This is example:
```python
from pygame import mixer

# Init mixer
mixer.init()

# Play mp3 file
mixer.Channel(0).play(mixer.Sound('sound.mp3'))

# Stop sound
if mixer.Channel(0).get_busy():
    mixer.Channel(0).stop()

# Stop and play
if mixer.Channel(0).get_busy():
    mixer.Channel(0).stop()
    mixer.Channel(0).play(mixer.Sound('sound.mp3'))
```

**Reply by glenn-jocher** [reply]
@dh031200 it sounds like you're experiencing overlapping audio because the new threads are being spawned for each detection without waiting for the previous audio to finish. To prevent this, you could implement a simple check to see if the previous sound is still playing before starting a new one.

One way to do this is to use a flag that indicates whether the sound is currently playing. You would set this flag to `True` when you start playing the sound and set it back to `False` when the sound finishes. You can modify the `play_audio` function to check this flag and only play the sound if it's not already playing.

Here's a conceptual example:

```python
from playsound import playsound
import threading

# Flag to indicate if the sound is currently playing
is_playing = False

def play_audio():
    global is_playing
    if not is_playing:
        is_playing = True
        playsound("welcome.wav")
        is_playing = False

for r in results:
    if 0 in r.boxes.cls:
        # Start the audio thread if the sound is not already playing
        if not is_playing:
            audio_thread = threading.Thread(target=play_audio)
            audio_thread.start()
```

Please note that this is a simplified example and may not be thread-safe. In a multi-threaded environment, you should use proper synchronization mechanisms like locks to prevent race conditions. However, this should give you an idea of how you can prevent the sound from playing multiple times simultaneously.
